injection/sharedmain/main.go

```diff
@@ -245,3 +287,89 @@ func flush(logger *zap.SugaredLogger) {
 	metrics.FlushExporter()
 }
+
+// ParseAndGetConfigOrDie parses the rest config flags and creates a client or
+// dies by calling log.Fatalf.
+func ParseAndGetConfigOrDie() *rest.Config {
+	var (
+		masterURL = flag.String("master", "",
+			"The address of the Kubernetes API server. Overrides any value in kubeconfig. Only required if out-of-cluster.")
+		kubeconfig = flag.String("kubeconfig", "",
+			"Path to a kubeconfig. Only required if out-of-cluster.")
+	)
+	flag.Parse()
+
+	cfg, err := GetConfig(*masterURL, *kubeconfig)
+	if err != nil {
+		log.Fatalf("Error building kubeconfig: %v", err)
+	}
+
+	return cfg
+}
+
+// MemStatsOrDie sets up reporting on Go memory usage every 30 seconds or dies
+// by calling log.Fatalf.
+func MemStatsOrDie(ctx context.Context) {
+	msp := metrics.NewMemStatsAll()
+	msp.Start(ctx, 30*time.Second)
+
+	if err := view.Register(msp.DefaultViews()...); err != nil {
+		log.Fatalf("Error exporting go memstats view: %v", err)
+	}
+}
+
+// SetupLoggerOrDie sets up the logger using the config from the given context
+// and returns a logger and atomic level, or dies by calling log.Fatalf.
+func SetupLoggerOrDie(ctx context.Context, component string) (*zap.SugaredLogger, zap.AtomicLevel) {
+	loggingConfig, err := GetLoggingConfig(ctx)
+	if err != nil {
+		log.Fatalf("Error reading/parsing logging configuration: %v", err)
+	}
+	logger, atomicLevel := logging.NewLoggerFromConfig(loggingConfig, component)
+	ctx = logging.WithLogger(ctx, logger)
+
+	return logger, atomicLevel
+}
+
+// SetupConfigMapWatchOrDie establishes a watch of the configmaps in the system
+// namespace that are labeled to be watched or dies by calling log.Fatalf.
+func SetupConfigMapWatchOrDie(ctx context.Context, logger *zap.SugaredLogger) *configmap.InformedWatcher {
+	kc := kubeclient.Get(ctx)
+	// Create ConfigMaps watcher with optional label-based filter.
+	var cmLabelReqs []labels.Requirement
+	if cmLabel := system.ResourceLabel(); cmLabel != "" {
+		req, err := configmap.FilterConfigByLabelExists(cmLabel)
+		if err != nil {
+			logger.With(zap.Error(err)).Fatalf("Failed to generate requirement for label %q")
+		}
+		logger.Infof("Setting up ConfigMap watcher with label selector %q", req)
+		cmLabelReqs = append(cmLabelReqs, *req)
+	}
+	// TODO(mattmoor): This should itself take a context and be injection-based.
+	cmw := configmap.NewInformedWatcher(kc, system.Namespace(), cmLabelReqs...)
+	return cmw
+}
+
+// WatchLoggingConfigOrDie establishes a watch of the logging config or dies by
+// calling log.Fatalf.
+func WatchLoggingConfigOrDie(ctx context.Context, cmw *configmap.InformedWatcher, logger *zap.SugaredLogger, atomicLevel zap.AtomicLevel, component string) {
+	if _, err := kubeclient.Get(ctx).CoreV1().ConfigMaps(system.Namespace()).Get(logging.ConfigMapName(),
+		metav1.GetOptions{}); err == nil {
+		cmw.Watch(logging.ConfigMapName(), logging.UpdateLevelFromConfigMap(logger, atomicLevel, component))
+	} else if !apierrors.IsNotFound(err) {
```

Comment:
What happens if it is not found? Seems like it should also fail? Or put another way, is there a reason not to die on any error?

Reply:
This is a copy of the code as it lives in the Main function today. This has been changed not to cause the component to die a while ago.

Reply:
Then I think at the very least we should clearly document what's going on. There are cases where this will NOT establish a watch nor die? Seems like if there's any reason (temporary can't reach), this will fail? Or maybe I'm just misunderstanding something :)

Reply:
See https://github.com/knative/pkg/pull/872 for the broader reasoning of that behavior. I agree we should add into the comment that this will not watch the logging config if it doesn't exist (but die on any other error)